File: go/protocols/mpp/client/session_consumer.go
// Kafka-style client helpers for metered session deliveries.
//
// SessionConsumer wraps an ActiveSession so applications can process delivered
// messages and call Ack/Commit instead of manually signing and posting
// vouchers. A failed commit never advances the local watermark, so the same
// directive can be retried safely.
package client

import (
	"context"
	"fmt"

	"github.com/solana-foundation/pay-kit/go/protocols/mpp/intents"
)

// CommitTransport sends a commit payload to the server and returns its receipt.
//
// HTTP clients, queues, and in-process tests all implement this. The directive
// is passed alongside the payload so transports can use CommitURL, Proof, or
// other routing hints without repeating them in the signed commit body.
type CommitTransport interface {
	Commit(ctx context.Context, directive intents.MeteringDirective, payload intents.CommitPayload) (intents.CommitReceipt, error)
}

// SessionConsumer is a client-side consumer for session-metered deliveries.
//
// SessionConsumer is not safe for concurrent use; the underlying ActiveSession
// watermark is advanced under Commit.
type SessionConsumer struct {
	// session is the wrapped ActiveSession; its cumulative watermark only
	// advances after the transport reports a successful commit.
	session *ActiveSession

	// transport posts signed commit payloads to the server and returns its
	// receipts.
	transport CommitTransport
}

// NewSessionConsumer wraps a session and a commit transport.
func NewSessionConsumer(session *ActiveSession, transport CommitTransport) *SessionConsumer {
	return &SessionConsumer{session: session, transport: transport}
}

// Session returns the wrapped session.
func (c *SessionConsumer) Session() *ActiveSession { return c.session }

// CommitDirective signs a voucher for the directive amount, sends it through
// the transport, and advances the local watermark only on success. It rejects
// directives whose session does not match, whose amount is not a valid base-unit
// integer, or whose amount is zero.
func (c *SessionConsumer) CommitDirective(ctx context.Context, directive intents.MeteringDirective) (intents.CommitReceipt, error) {
	if err := c.validateDirective(directive); err != nil {
		return intents.CommitReceipt{}, err
	}
	amount, err := directive.AmountBaseUnits()
	if err != nil {
		return intents.CommitReceipt{}, err
	}
	if amount == 0 {
		return intents.CommitReceipt{}, fmt.Errorf("metered delivery amount must be greater than zero")
	}

	voucher, err := c.session.PrepareIncrement(amount)
	if err != nil {
		return intents.CommitReceipt{}, err
	}
	payload := intents.CommitPayload{DeliveryID: directive.DeliveryID, Voucher: voucher}

	receipt, err := c.transport.Commit(ctx, directive, payload)
	if err != nil {
		return intents.CommitReceipt{}, err
	}
	// A replayed receipt means the server already settled this delivery, so its
	// Cumulative is the authoritative settled position. Recording the freshly
	// prepared (higher) voucher would push the local watermark past the server's
	// state and let a later close sign for more than was agreed; skipping it
	// entirely would instead leave the watermark behind the server when the
	// original response was lost, so the next delivery signs a non-monotonic
	// cumulative. Reconcile to the receipt cumulative on replay (never
	// regressing); record the voucher on a fresh committed receipt.
	switch receipt.Status {
	case intents.CommitStatusReplayed:
		settled, perr := parseCumulative(receipt.Cumulative)
		if perr != nil {
			return intents.CommitReceipt{}, fmt.Errorf("invalid replayed receipt cumulative: %w", perr)
		}
		// The server is untrusted: clamp to the voucher just prepared in this
		// call. An honest lost-response replay settles at or below it (the
		// session is single-threaded), so a server reporting a higher cumulative
		// cannot push the watermark past what the client actually signed —
		// otherwise the next voucher would over-authorize up to the deposit.
		prepared, perr := parseCumulative(voucher.Data.Cumulative)
		if perr != nil {
			return intents.CommitReceipt{}, fmt.Errorf("invalid prepared voucher cumulative: %w", perr)
		}
		if settled > prepared {
			settled = prepared
		}
		c.session.ReconcileSettled(settled)
	case intents.CommitStatusCommitted:
		if err := c.session.RecordVoucher(voucher); err != nil {
			return intents.CommitReceipt{}, err
		}
	default:
		// A malformed or unknown status must not advance local state.
		return intents.CommitReceipt{}, fmt.Errorf("unexpected commit receipt status: %q", receipt.Status)
	}
	return receipt, nil
}

func (c *SessionConsumer) validateDirective(directive intents.MeteringDirective) error {
	channelID := c.session.ChannelIDString()
	if directive.SessionID != channelID {
		return fmt.Errorf(
			"metered delivery session %s does not match active session %s", directive.SessionID, channelID)
	}
	return nil
}

// Accept validates an envelope and returns a delivery handle exposing Ack and
// Commit. The directive is validated up front so a mismatched session is
// rejected before the application processes the payload.
func Accept[P any](c *SessionConsumer, envelope intents.MeteredEnvelope[P]) (*MeteredDelivery[P], error) {
	if err := c.validateDirective(envelope.Metering); err != nil {
		return nil, err
	}
	return &MeteredDelivery[P]{
		consumer: c,
		payload:  envelope.Payload,
		metering: envelope.Metering,
	}, nil
}

// MeteredDelivery is a delivered payload paired with its metering directive.
// Call Ack (or its Commit alias) after the application has processed Payload.
type MeteredDelivery[P any] struct {
	// consumer is the consumer that accepted the delivery; Ack commits the
	// directive amount through it.
	consumer *SessionConsumer

	// payload is the delivered application payload.
	payload P

	// metering is the directive pricing this delivery; Ack signs a voucher
	// for its amount.
	metering intents.MeteringDirective
}

// Payload returns the delivered payload.
func (d *MeteredDelivery[P]) Payload() P { return d.payload }

// Metering returns the metering directive that accompanied the payload.
func (d *MeteredDelivery[P]) Metering() intents.MeteringDirective { return d.metering }

// Ack signs and commits a voucher for the directive amount.
func (d *MeteredDelivery[P]) Ack(ctx context.Context) (intents.CommitReceipt, error) {
	return d.consumer.CommitDirective(ctx, d.metering)
}

// Commit is an alias for Ack.
func (d *MeteredDelivery[P]) Commit(ctx context.Context) (intents.CommitReceipt, error) {
	return d.Ack(ctx)
}

// IntoParts returns the payload and metering directive without committing.
func (d *MeteredDelivery[P]) IntoParts() (P, intents.MeteringDirective) {
	return d.payload, d.metering
}

File: python/src/solana_pay_kit/protocols/mpp/client/session_consumer.py
"""Kafka-style client helpers for metered session deliveries.

:class:`SessionConsumer` wraps an :class:`~solana_pay_kit.protocols.mpp.client.session.ActiveSession`
so applications can process delivered messages and call ``ack``/``commit``
instead of manually signing and posting vouchers. A failed commit never
advances the local watermark, so the same directive can be retried safely.
"""

from __future__ import annotations

from typing import Any, Generic, Protocol, TypeVar, runtime_checkable

from solana_pay_kit.protocols.mpp.client.session import ActiveSession
from solana_pay_kit.protocols.mpp.intents.session import (
    CommitPayload,
    CommitReceipt,
    MeteredEnvelope,
    MeteringDirective,
)

__all__ = [
    "CommitTransport",
    "SessionConsumer",
    "MeteredDelivery",
]

P = TypeVar("P")


@runtime_checkable
class CommitTransport(Protocol):
    """Transport used by :class:`SessionConsumer` to send commit payloads.

    HTTP clients, queues, and in-process tests can all implement this. The
    directive is passed alongside the payload so transports can use
    ``commit_url``, ``proof``, or other routing hints without those fields being
    repeated in the signed commit body.

    The ``commit`` method takes the directive and the :class:`CommitPayload` and
    returns a :class:`CommitReceipt`.
    """

    def commit(self, directive: MeteringDirective, payload: CommitPayload) -> CommitReceipt:
        """Send ``payload`` for ``directive`` and return the server receipt."""
        ...


class SessionConsumer:
    """Client-side consumer for session-metered deliveries.

    Not safe for concurrent use; the underlying :class:`ActiveSession` watermark
    is advanced inside :meth:`commit_directive`.
    """

    def __init__(self, session: ActiveSession, transport: CommitTransport) -> None:
        """Wrap a session and a commit transport."""
        self._session = session
        self._transport = transport

    @property
    def session(self) -> ActiveSession:
        """The wrapped session."""
        return self._session

    @property
    def transport(self) -> CommitTransport:
        """The wrapped commit transport."""
        return self._transport

    def accept(self, envelope: MeteredEnvelope) -> MeteredDelivery[Any]:
        """Validate an envelope and return a delivery handle with ``ack``/``commit``.

        The directive is validated up front so a mismatched session is rejected
        before the application processes the payload.
        """
        self._validate_directive(envelope.metering)
        return MeteredDelivery(consumer=self, payload=envelope.payload, metering=envelope.metering)

    def commit_directive(self, directive: MeteringDirective) -> CommitReceipt:
        """Sign a voucher for the directive amount, send it, and advance the
        local watermark only on success.

        Rejects directives whose session does not match, whose amount is not a
        valid base-unit integer, or whose amount is zero. The prepare/record
        split makes a failed commit safe to retry without double-counting: the
        voucher is prepared (no watermark advance), sent, and recorded once the
        transport returns a receipt. The prepared voucher is recorded for both
        ``committed`` and ``replayed`` receipts: the server's deliveryId dedupe
        keeps the settled amount authoritative on its side, and the locally
        signed cumulative stays the client watermark so subsequent vouchers
        remain monotonic.
        """
        self._validate_directive(directive)
        amount = directive.amount_base_units()
        if amount == 0:
            raise ValueError("metered delivery amount must be greater than zero")

        voucher = self._session.prepare_increment(amount)
        payload = CommitPayload(delivery_id=directive.delivery_id, voucher=voucher)

        receipt = self._transport.commit(directive, payload)
        if receipt.status == "replayed":
            # The server already settled this delivery; its cumulative is the
            # authoritative settled position. Clamp to the just-prepared voucher
            # so an untrusted server cannot push the watermark past what we
            # signed. Reconcile (never regress) instead of recording the voucher.
            prepared = int(payload.voucher.data.cumulative)
            self._session.reconcile_settled(min(receipt.cumulative_base_units(), prepared))
        else:
            self._session.record_voucher(payload.voucher)
        return receipt

    def _validate_directive(self, directive: MeteringDirective) -> None:
        channel_id = self._session.channel_id_string
        if directive.session_id != channel_id:
            raise ValueError(
                f"metered delivery session {directive.session_id} does not match active session {channel_id}"
            )


class MeteredDelivery(Generic[P]):
    """A delivered payload paired with its metering directive.

    Call :meth:`ack` (or its :meth:`commit` alias) after the application has
    processed :attr:`payload`.
    """

    def __init__(self, consumer: SessionConsumer, payload: P, metering: MeteringDirective) -> None:
        self._consumer = consumer
        self._payload = payload
        self._metering = metering

    @property
    def payload(self) -> P:
        """The delivered payload."""
        return self._payload

    @property
    def metering(self) -> MeteringDirective:
        """The metering directive that accompanied the payload."""
        return self._metering

    def ack(self) -> CommitReceipt:
        """Sign and commit a voucher for the directive amount."""
        return self._consumer.commit_directive(self._metering)

    def commit(self) -> CommitReceipt:
        """Alias for :meth:`ack`."""
        return self.ack()

    def into_parts(self) -> tuple[P, MeteringDirective]:
        """Return the payload and metering directive without committing."""
        return self._payload, self._metering
